cmd/tailscale/cli/serve_dev.go

```diff
@@ -134,14 +135,104 @@ func (e *serveEnv) runServeDev(funnel bool) execFunc {
 }
 
 func (e *serveEnv) streamServe(ctx context.Context, req ipn.ServeStreamRequest) error {
-	stream, err := e.lc.StreamServe(ctx, req)
+	watcher, err := e.lc.WatchIPNBus(ctx, ipn.NotifyInitialState|ipn.NotifyServeRequest)
 	if err != nil {
 		return err
 	}
-	defer stream.Close()
+	defer watcher.Close()
+	n, err := watcher.Next()
+	if err != nil {
+		return err
+	}
+	sessionID := n.SessionID
+	if sessionID == "" {
+		return errors.New("missing session id")
+	}
+	sc, err := e.lc.GetServeConfig(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting serve config: %w", err)
+	}
+	if sc == nil {
+		sc = &ipn.ServeConfig{}
+	}
+	setHandler(sc, req, sessionID)
+	err = e.lc.SetServeConfig(ctx, sc)
+	if err != nil {
+		return fmt.Errorf("error setting serve config: %w", err)
+	}
+
+	fmt.Fprintf(os.Stderr, "Funnel started on \"https://%s\".\n", strings.TrimSuffix(string(req.HostPort), ":443"))
+	fmt.Fprintf(os.Stderr, "Press Ctrl-C to stop Funnel.\n\n")
+
+	for {
+		n, err := watcher.Next()
+		if err != nil {
+			return fmt.Errorf("error calling next: %w", err)
+		}
+		if n.FunnelRequestLog == nil || n.FunnelRequestLog.SessionID != sessionID {
+			continue
+		}
+		bts, _ := json.Marshal(n.FunnelRequestLog)
+		fmt.Printf("%s\n", bts)
+	}
+}
+
+func setHandler2(sc *ipn.ServeConfig, req ipn.ServeStreamRequest, sessionID string) {
```

Comment:
2? what's 1? docs?

Reply:
@bradfitz that was me debugging stuff. deleting.